### ISSUE TITLE
add_smart_highlight_annotations does not work for paragraphs

### ISSUE THREAD

**Issue by danieluhricek** [lucasrla/remarks#52]
[add_smart_highlight_annotations](https://github.com/lucasrla/remarks/blob/master/remarks/conversion/drawing.py#L176) does not work when having full paragraph smart highlights. 

For example:
Highlighted text is a 5 lines paragraph. `hl["text"]` contains whole 5 lines of text and `page.search_for(hl["text"], quads=True)` returns no matches. `page.search_for` matches only on text max length of 1 line.

Example highlights json:
```
{
    "highlights": [
        [
            {
                "color": 3,
                "length": 454,
                "rects": [
                    {
                        "height": 27.12016196910554,
                        "width": 1009.0730312247251,
                        "x": 140.4741632434343,
                        "y": 599.9028268116126
                    },
                    {
                        "height": 27.974953197517607,
                        "width": 1029.1417209793326,
                        "x": 114.89519231887243,
                        "y": 630.2660855855902
                    },
                    {
                        "height": 27.12016196910554,
                        "width": 1033.5790201634973,
                        "x": 115.73704027993699,
                        "y": 662.0772446970841
                    },
                    {
                        "height": 27.12016196910554,
                        "width": 1034.2260248394723,
                        "x": 115.55571551852154,
                        "y": 693.1527501490382
                    },
                    {
                        "height": 27.12016196910554,
                        "width": 807.2974610806432,
                        "x": 115.11536671875547,
                        "y": 724.2282556009923
                    }
                ],
                "start": 819,
                "text": "This work is motivated by our own environment at the European Organization for NuclearResearch (CERN), a large multinational research organization where thousands of visiting researchers bring IoT devices into the network for research purposes. Devices includes networkedoscilloscopes, storage, printers, cameras, all kinds of networked sensors that researchers can easilyintegrate into the network with the help of a bring-your-own-device policy."
            }
        ]
    ]
}
```

Would it be possible to fix it using provided `rects` directly in highlights json (instead of `page.search_for`)? Do you know whats the relation between `rects` dimensions and quads in pymupdf?

---

The rect[0] from highlights json would be (x0, y0, x1, y1):
```
(140.4741632434343, 599.9028268116126, 1149.5471944681594, 627.0229887807182)
```

Bbox highlighting the same line in pymupdf (got from `page.get_text("dict")`):
```
(55.73711013793945, 228.21629333496094, 440.51470947265625, 242.69195556640625)
```

In this case it would somehow fix by scaling down rect[0] coordinates by numbers in range (2.52, 2.63).

---

The rect[1] from highlights json:
```
(114.89519231887243, 630.2660855855902, 1144.036913298205, 658.2410387831078)
```

and corresponding bbox:
```
(45.77450942993164, 240.16844177246094, 441.9892578125, 254.64410400390625)
```

again possible fixing by scaling down by similar range? The scaling factor for individual coordinates (x0, y0, x1, y1) seems pretty constant (delta 0.01).

---

If you don't want to go this way, simple fix would be any of:

1. determining line length and searching for substrings of `hl["text"]`
2. bruteforcing the longest matchable prefix from `hl["text"]`, removing matched text - repeat until `hl["text"]` empty
3. iterating `page.get_text("dict")["blocks"]` and their `["lines"]`

---

Do you have any preference how it should be fixed?

**Comment by danieluhricek** [comment]
I'm currently using this quick fix:

```python
.
.
.
for hl in hl_list:
        # https://pymupdf.readthedocs.io/en/latest/page.html#Page.add_highlight_annot
        quads = page.search_for(hl["text"], quads=True)

        if not quads:
            logging.warning(f"- Didn't find text in page. Using hl['rects'].")

            k = 2.6
            padding = 2

            quads = [
                fitz.Rect(
                    (rect["x"] - padding) / k,
                    (rect["y"] - padding) / k,
                    (rect["x"] + rect["width"] + padding) / k,
                    (rect["y"] + rect["height"] + padding) / k,
                )
                for rect in hl["rects"]
            ]

        annot = page.add_highlight_annot(quads)
.
.
.
```

**Comment by lucasrla** [comment]
That's an interesting quick fix, @danieluhricek. I'll test it locally on my files in the upcoming days. 

If it fares well, I will push it to the master branch.

If not, I will investigate both `page.search_for` limitations (it's been a while and I don't remember exactly how it behaves) and your third suggestion (iterating over `page.get_text("dict")["blocks"]` lines).

Thanks!

**Comment by lucasrla** [comment]
I've just investigated your issue and have now pushed a fix ([55eb1ad](https://github.com/lucasrla/remarks/commit/12e1bde4265cc5254a14e814f9a4af4269e6ea3d)) based on some of the ideas you brought up. Thanks!

For the details, read on.

***

First of all, `page.search_for` does find multi-line text on PDFs. Your issue is not PyMuPDF's fault, it's actually reMarkable's!

`page.search_for` is correct to return an empty array of quads for your query. The actual problem is that we're asking it to search for the wrong text (which it correctly never finds). We obtain that text from reMarkable's own smart highlights, but the string often is poorly constructed.

Take a closer look at the JSON content you provided above. reMarkable makes mistakes when concatenating/joining some of the lines (i.e., "NuclearResearch", "networkedoscilloscopes", "easilyintegrate"):

```json
"text": "This work is motivated by our own environment at the European Organization for NuclearResearch (CERN), a large multinational research organization where thousands of visiting researchers bring IoT devices into the network for research purposes. Devices includes networkedoscilloscopes, storage, printers, cameras, all kinds of networked sensors that researchers can easilyintegrate into the network with the help of a bring-your-own-device policy."
```

In fact, I found your PDF on Google and could reproduce your exact issue within Python interactive console:

1. Download a PDF copy of "[Risk Prediction of IoT Devices Based on Vulnerability Analysis](https://dl.acm.org/doi/10.1145/3510360)" and save it to `demo/_input/3510360.pdf`
2. cd to your local `remarks` dir
3. Run `python`
4. Run the following:

```python
Python 3.10.9 (main, Jan 11 2023, 17:17:53) [Clang 11.1.0 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import fitz

>>> doc = fitz.open("demo/_input/3510360.pdf")

>>> doc.page_count
36

>>> page = doc.load_page(1)

>>> page.search_for("This work is motivated by our own environment at the European Organization for NuclearResearch", quads=True)
[]

>>> page.search_for("This work is motivated by our own environment at the European Organization for Nuclear Research (CERN)", quads=True)
[Quad(Point(55.73711013793945, 228.21629333496094), Point(279.189208984375, 228.21629333496094), Point(55.73711013793945, 241.984619140625), Point(279.189208984375, 241.984619140625)), Quad(Point(279.189208984375, 229.3221435546875), Point(440.51470947265625, 229.3221435546875), Point(279.189208984375, 242.69195556640625), Point(440.51470947265625, 242.69195556640625)), Quad(Point(45.77450942993164, 241.2742919921875), Point(123.95999908447266, 241.2742919921875), Point(45.77450942993164, 254.64410400390625), Point(123.95999908447266, 254.64410400390625))]
```

The other point worth documenting here is the mechanics behind the scaling factor. In your case, it's 2.6 indeed, but it varies depending on PDF page dimensions.

Continuing on the same Python's REPL I started above:

```python
>>> page.rect
Rect(0.0, 0.0, 486.0, 720.0)

# Now, recall that the reMarkable device dimensions are
# RM_WIDTH = 1404
# RM_HEIGHT = 1872

# And here's why you got to 2.6
>>> 1872/720
2.6
```

Given that remarks already calculates the appropriate scaling factor for each page in `rescale_given_device_aspect_ratio`, the fix I pushed in my commit uses that variable instead of your `k`.
